shiny/render/_data_frame_utils/_tbl_data.py

```diff
@@ -43,356 +38,299 @@
-            if cols is None
-            else [
-                (
-                    # Send in an array of size 1 and retrieve the first element
-                    data.columns.get_indexer_for(  # pyright: ignore[reportUnknownMemberType]
-                        [col]
-                    )[
-                        0
-                    ]
-                    if isinstance(col, str)
-                    else col
-                )
-                for col in cols
-            ]
-        )
-
-        # Force list when using a non-None value for pandas compatibility
-        indx_rows = list(rows) if rows is not None else slice(None)
-
-        return data.iloc[indx_rows, indx_cols]
-
-
-@subset_frame.register
-def _(
-    data: PlDataFrame,
-    *,
-    rows: RowsList = None,
-    cols: ColsList = None,
-) -> PlDataFrame:
-    indx_cols = (
-        [col if isinstance(col, str) else data.columns[col] for col in cols]
-        if cols is not None
-        else slice(None)
-    )
-    indx_rows = rows if rows is not None else slice(None)
-    return data[indx_rows, indx_cols]
-
-
-# get_frame_cell -----------------------------------------------------------------------
-
-
-@singledispatch
-def get_frame_cell(data: DataFrameLike, row: int, col: int) -> Any:
-    raise TypeError(f"Unsupported type: {type(data)}")
-
-
-@get_frame_cell.register
-def _(data: PdDataFrame, row: int, col: int) -> Any:
-    return (
-        data.iat[  # pyright: ignore[reportUnknownMemberType, reportUnknownVariableType]
-            row, col
-        ]
-    )
+    if cols is None:
+        if rows is None:
+            return data
+        else:
+            # This feels like it should be `data[rows, :]` but that doesn't work for polars
+            # https://github.com/narwhals-dev/narwhals/issues/989
+            # Must use `list(rows)` as tuples are not supported
+            # https://github.com/narwhals-dev/narwhals/issues/990
+            return data[list(rows), data.columns]
+    else:
+        # `cols` is not None
+        col_names = [data.columns[col] if isinstance(col, int) else col for col in cols]
+        # If len(cols) == 0, then we should return an empty DataFrame
+        # Currently this is broken when backed by pandas.
+        # https://github.com/narwhals-dev/narwhals/issues/989
+        if len(col_names) == 0:
+            # Return a DataFrame with no rows or columns
+            # If there are no columns, there are no Series objects to support any rows
+            return data[[], []]
+
+        if rows is None:
+            return data[:, col_names]
+        else:
+            # Be sure rows is a list, not a tuple. Tuple
+            return data[list(rows), col_names]
 
 
-@get_frame_cell.register
-def _(data: PlDataFrame, row: int, col: int) -> Any:
-    return data[row, col]
+# # get_frame_cell -----------------------------------------------------------------------
+def get_frame_cell(data: DataFrame[Any], row: int, col: int) -> Any:
```

Comment:
Remove the very simple functions and just inline. We're not afraid to tightly bound to Narwhals' (stable) API